grapevine/src/subscriber/sequencer.rs

```diff
@@ -142,56 +155,62 @@ impl<S: StateSync> MessageSequencer<S> {
-                            Ok(response) => {
-                                let mut stream = response.into_inner();
-                                // TODO: err
-                                let msg = stream.next().await.ok_or("Empty stream")??;
-                                self.want_sequence += 1;
-                                self.last_deadline = now + self.sequence_timeout;
-                                while let Some(Ok(msg)) = stream.next().await {
-                                    trace!("Pushing {} to heap from stream", msg.metadata.sequence);
-                                    self.heap.push(Reverse(msg));
-                                }
-                                Ok(SubscriberStream::Message(Self::try_parse(&msg)?))
-                            }
-                            Err(err) => Err(Box::new(err)),
-                        };
-                    } else if missing <= 1 {
+                    if single_missing {
                         debug!("Requesting message {}", self.want_sequence);
                         let msg = self
                             .recovery_client
                             .get_message(SequenceRequest {
                                 sequence_id: self.want_sequence,
                             })
                             .await?;
+                        let msg = ReceivedMessage::new(msg.into_inner(), Origin::Missing);
+
+                        if msg.raw.metadata.sequence != self.want_sequence {
+                            return Err(Box::new(SubscriberStreamError::InvalidSequence));
+                        }
+
                         self.want_sequence += 1;
                         self.last_deadline = now + self.sequence_timeout;
-                        return Ok(SubscriberStream::Message(Self::try_parse(
-                            &msg.into_inner(),
-                        )?));
+                        return Ok(Self::try_parse(&msg)?);
                     } else {
-                        // recover multiple
-                        // TODO: log messages
-                        debug!("Requesting messages");
-                        todo!()
-                    };
+                        // restream from want_sequence
+                        debug!("Requesting sync from {}", self.want_sequence);
+
+                        let response = self
+                            .recovery_client
+                            .stream_from(Request::new(SequenceRequest {
+                                sequence_id: self.want_sequence,
+                            }))
+                            .await?;
+
+                        let mut stream = response.into_inner();
+
+                        // if don't clean heap then there will be duplicates
+                        // and origin (udp/sync) will be mixed
+                        // other solution is to implement Ord which compare tuples (seq_id, origin)
+                        self.heap.clear();
```

Comment:
Good point, although this might make syncing quite difficult for high-throughput streams.

It looks like you push all syncing messages on to the heap below, but that shouldn't be necessary. It's also not  memory efficient.

We should be able to process all sync messages as they arrive, and then look back at the heap with out new `want_sequence`, discarding any old messages (therefore all duplicates).

Seeing as this looks to work currently we should do that in a new PR.

Reply:
okay